src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -2304,14 +2333,189 @@ private void CheckIndexerSignatureCollisions(
+                        }
+                    }
+                }
+
+                return (membersByName, membersUnordered);
+            }
+
+            static ImmutableArray<Symbol> concatMembers(ImmutableArray<Symbol> existingMembers, NamedTypeSymbol extension, ImmutableArray<Symbol> newMembers, ref ImmutableArray<Symbol> membersUnordered)
+            {
+                Debug.Assert(!newMembers.IsEmpty);
+
+                if (extension.IsDefinition)
+                {
+                    Debug.Assert(newMembers.All(static (m, membersUnordered) => membersUnordered.Contains(m), membersUnordered));
+                    return existingMembers.Concat(newMembers);
+                }
+
+                var membersBuilder = ArrayBuilder<Symbol>.GetInstance(existingMembers.Length + newMembers.Length);
+                var membersUnorderedBuilder = ArrayBuilder<Symbol>.GetInstance(membersUnordered.Length + newMembers.Length);
+
+                membersBuilder.AddRange(existingMembers);
+                membersUnorderedBuilder.AddRange(membersUnordered);
+
+                foreach (var member in newMembers)
+                {
+                    Symbol toAdd = member.SymbolAsMember(extension);
+                    membersBuilder.Add(toAdd);
+                    membersUnorderedBuilder.Add(toAdd);
+                }
+
+                membersUnordered = membersUnorderedBuilder.ToImmutableAndFree();
+                return membersBuilder.ToImmutableAndFree();
+            }
+        }
+
+        private readonly struct ExtensionGroupingKey : IEquatable<ExtensionGroupingKey>
+        {
+            public readonly int ExtensionArity;
+            public readonly TypeSymbol ReceiverType;
+
+            public ExtensionGroupingKey(NamedTypeSymbol extension)
+            {
+                ExtensionArity = extension.Arity;
+
+                if (extension.Arity != 0)
+                {
+                    extension = extension.Construct(IndexedTypeParameterSymbol.Take(extension.Arity));
+                }
+
+                if (extension.ExtensionParameter is { } receiverParameter)
+                {
+                    ReceiverType = receiverParameter.Type;
+                }
+                else
+                {
+                    ReceiverType = ErrorTypeSymbol.UnknownResultType;
+                }
+            }
+
+            public bool Equals(ExtensionGroupingKey other)
+            {
+                return ExtensionArity == other.ExtensionArity &&
+                       ReceiverType.Equals(other.ReceiverType, TypeCompareKind.AllIgnoreOptions);
+            }
+
+            public override bool Equals([NotNullWhen(true)] object? obj)
+            {
+                Debug.Assert(false); // Usage of this method is unexpected
+                return Equals((ExtensionGroupingKey)obj!);
+            }
+
+            public override int GetHashCode()
+            {
+                return ReceiverType.GetHashCode();
```

Comment:
Shouldn't `TypeCompareKind.AllIgnoreOptions` be taken into consideration here as well? Otherwise it might happen that two `ExtensionGroupingKey`s that are equal by `Equals` have different hash code (because they have nullability annotations differences for example).

Reply:
> Shouldn't ```TypeCompareKind.AllIgnoreOptions``` be taken into consideration here as well?

It should be and it is taken into consideration. Implementations of `GetHashCode` for symbols are supposed to ignore any differences that could be ignored during symbol comparison. That significantly simplifies code in here and in ```SymbolEqualityComparer```, for example.